vertx-web-client/src/test/java/io/vertx/ext/web/client/WebClientTest.java

```diff
@@ -1104,4 +1106,22 @@ public void testStreamHttpServerRequest() throws Exception {
       server2.close();
     }
   }
+
+  @Test
+  public void testClientErrors() {
+    WebClientOptions wcOptions = new WebClientOptions().setUserAgent("Vertx GitHub Client").setDefaultHost("api.github.com").setDefaultPort(443).setSsl(true);
+    WebClient client = WebClient.create(vertx, wcOptions);
+    HttpRequest<GHUser> request = client.get("/users/" + UUID.randomUUID().toString()).as(BodyCodec.json(GHUser.class));
```

Comment:
I'm forcing a `404` response by the server in order to receive this JSON payload:

```
{
  "message": "Not Found",
  "documentation_url": "https://developer.github.com/v3/users/#get-a-single-user"
}
```

Reply:
should we try to do pojo mapping only when the response is a 200 ?

Reply:
@vietj would this restriction only apply to Json POJO codec?

Reply:
This would work, right?

```java
HttpRequest<GHUser> request = client.get("/users/" + UUID.randomUUID().toString());
request.send(ar -> {
  if (ar.suceeded() {
    HttpResponse<Buffer> response = ar.result();
    if(response.statusCode()==200) {
      GHUser user = response.bodyAsJson(GHUser.class);
    } else {
      JsonObject json = response.bodyAsJsonObject(); // or create a GHError object and do as with GHUser
    }
  } else {
    // boom
  }
});
```

Reply:
@vietj 

> Should we try to do pojo mapping only when the response is a `200`?

I don't think so because we could receive other _positives_ responses, for example, `201` when we want to create an entity by using a `PUT` method.

Reply:
@tsegismont 

```java
HttpRequest<Buffer> request = client.get("/users/" + UUID.randomUUID().toString());
request.send(ar -> {
  if (ar.suceeded() {
    HttpResponse<Buffer> response = ar.result();
    if(response.statusCode()==200) {
      GHUser user = response.bodyAsJson(GHUser.class);
    } else {
      JsonObject json = response.bodyAsJsonObject(); // or create a GHError object and do as with GHUser
    }
  } else {
    // boom
  }
});
```

We could do it but I see these points:

* You have to ask for other HTTP Codes, for example `201` because many REST API returns the entity created after a `PUT` method. ([rf](https://developer.github.com/v3/repos/#response-2).)
* Who uses the client needs to navigate for this `JsonObject` because the structure can be different depending of the request.
* On snippet of code we are assuming that `BodyCodec` (and other custom Codec for processing collection of entities) has been overridden because we will find a `RuntimeException` while process the payload (aka `body`).

Do my comments make sense for you?

Reply:
>You have to ask for other HTTP Codes, for example 201 because many REST API returns the entity created after a PUT method. (rf.)

Yes. As a user, you must know the contract of the API you're working with.

> Who uses the client needs to navigate for this JsonObject because the structure can be different depending of the request.

Are you talking about the `JsonObject` in case status code is not 200 ? Then as commented previously it's possible to get it as a custom `GHError` class.

>On snippet of code we are assuming that BodyCodec (and other custom Codec for processing collection of entities) has been overridden because we will find a RuntimeException while process the payload (aka body).

Not sure what you mean

Reply:
@tsegismont 

> Are you talking about the JsonObject in case status code is not 200 ? Then as commented previously it's possible to get it as a custom GHError class.

How can you get a `GHError` when you Codec is waiting for `GHUser`?

> Not sure what you mean

I'll add more test in order to reduce all scenarios where I'm seeing some improvements on this web client implementation.

Reply:
>How can you get a `GHError` when you Codec is waiting for `GHUser`?

@recena take a look at the snippet again, I did not specific and codec. That's why the body can be converted to anything in the response handler.

Reply:
@tsegismont It is true. If we do that, we are delegating too much responsibility to the user (IMHO). He is waiting for something more usable (UX).

Reply:
I need to think about which changes would be needed on `web-client` so this ext was more flexible.

Reply:
I think what we should have is something like expect() that restricts the expected returned code otherwise fails as an exception so there is no mapping.

```
client.get("/foo").expect(200).send(...)
```

if it does not get a 200 (or perhaps 2xx) then it fails with an exception , for instance 404 with a NotFoundException,etc....

Reply:
so in short it makes an interpretation of the response status code

Reply:
> If we do that, we are delegating too much responsibility to the user

I fail to see how this is a matter of responsibility to the user. In the first example the user provides a `BodyCodec` before sending the request, in the second a target type after receiving the response.

That said, the Web Client may defer the body conversion until the user actually requires the content. Then the exception would not happen because the user would still check the response code before retrieving the body.